xds/src/main/java/io/grpc/xds/internal/certprovider/CertificateProviderStore.java

```diff
@@ -0,0 +1,244 @@
+    }
+  }
+
+  @VisibleForTesting
+  CertificateProviderStore(CertificateProviderRegistry certificateProviderRegistry) {
+    this.certificateProviderRegistry = certificateProviderRegistry;
+    certProviderMap = new ReferenceCountingMap<>(new CertProviderFactory());
+  }
+
+  @VisibleForTesting
+  final class HandleImpl implements Handle {
+    private final CertProviderKey key;
+    private final Watcher watcher;
+    @VisibleForTesting
+    final CertificateProvider certProvider;
+
+    private HandleImpl(CertProviderKey key, Watcher watcher, CertificateProvider certProvider) {
+      this.key = key;
+      this.watcher = watcher;
+      this.certProvider = certProvider;
+    }
+
+    @Override
+    public synchronized void close() {
+      DistributorWatcher distWatcher = (DistributorWatcher)certProvider.getWatcher();
+      distWatcher.removeWatcher(watcher);
+      certProviderMap.release(key, certProvider);
+    }
+  }
+
+  /**
+   * Creates or retrieves a {@link CertificateProvider} instance, increments its ref-count and
+   * registers the watcher passed. Returns a {@link Handle} that can be {@link Handle#close()}d when
+   * the instance is no longer needed by the caller.
+   *
+   * @param notifyCertUpdates when true, the caller is interested in identity cert updates. When
+   *     false, the caller cannot depend on receiving the {@link Watcher#updateCertificate}
+   *     callbacks but may still receive these callbacks which should be ignored.
+   * @throws IllegalArgumentException in case of errors in processing config or the plugin is
+   *     incapable of sending cert updates when notifyCertUpdates is true.
+   * @throws UnsupportedOperationException if the plugin is incapable of sending cert updates when
+   *     notifyCertUpdates is true.
+   */
+  public synchronized Handle createOrGetProvider(
+      String certName,
+      String pluginName,
+      Object config,
+      Watcher watcher,
+      boolean notifyCertUpdates) {
+    if (!notifyCertUpdates) {
+      // we try to get a provider first for notifyCertUpdates==true always
+      try {
+        return createProviderHelper(certName, pluginName, config, watcher, true);
+      } catch (UnsupportedOperationException uoe) {
+        // ignore & log exception and fall thru to create a provider with actual value
+        logger.log(Level.FINE, "Trying to get provider for notifyCertUpdates==true", uoe);
+      }
+    }
+    return createProviderHelper(certName, pluginName, config, watcher, notifyCertUpdates);
+  }
+
+  private synchronized Handle createProviderHelper(
+      String certName,
+      String pluginName,
+      Object config,
+      Watcher watcher,
+      boolean notifyCertUpdates) {
+    CertProviderKey key = new CertProviderKey(certName, pluginName, notifyCertUpdates, config);
+    CertificateProvider provider = certProviderMap.get(key);
+    DistributorWatcher distWatcher = (DistributorWatcher) provider.getWatcher();
```

Comment:
This casting is not obvious. Why about defining `CertificateProvider` with a generic type for watcher:

```java
public abstract class CertificateProvider<T extends Watcher> implements Closeable {
  final T watcher;
  public T getWatcher() {
    return watcher;
   }
}
```

Reply:
Good point about the casting. But I realized a generic is not needed. The `CertificateProvider` always takes a `DistributorWatcher` so I moved the inner class here and removed the casting in all those places.